reinforcement_learning/environment/dialogue_env_moviebot.py

```diff
@@ -0,0 +1,361 @@
+            self._save_dialogue_history()
+
+        # 8. Get the next state
+        dialogue_state = (
+            self.agent.dialogue_manager.dialogue_state_tracker.get_state()
+        )
+        observation = NeuralDialoguePolicy.build_input(
+            dialogue_state,
+            b_use_intents=self.b_use_intents,
+            user_intents=user_intents,
+            agent_intents=agent_intents,
+        )
+
+        # 9. Additional information
+        info.update(
+            {
+                "reward": reward,
+                "agent_dialogue_acts": agent_dacts,
+            }
+        )
+
+        return observation, reward, done, truncated, info
+
+    def get_bonus_reward(self) -> Tuple[float, bool]:
+        """Returns the bonus reward and if the dialogue was successful.
+
+        The reward is computed as follows:
+        - If the user accepted one recommendation: +100
+        - If the agent made an offer but the user didn't accept one: -50
+        - If the agent didn't make a recommendation: -100
+        """
+        reward = 0.0
+        success = self.accepted_recommendation()
+
+        dialogue_state = (
+            self.agent.dialogue_manager.dialogue_state_tracker.get_state()
+        )
+
+        if success:
+            reward = 100.0
+        elif dialogue_state.agent_made_offer:
+            reward = -50.0
+        else:
+            reward = -100.0
+
+        return reward, success
+
+    def accepted_recommendation(self) -> bool:
+        """Returns True if the user accepted one recommendation."""
+        b_accept = False
+        dialogue_state = (
+            self.agent.dialogue_manager.dialogue_state_tracker.get_state()
+        )
+        for _, choices in dialogue_state.movies_recommended.items():
+            if "accept" in choices:
+                b_accept = True
+                break
+        return b_accept
+
+    def _save_dialogue_history(self) -> None:
+        """Saves the dialogue history.
+
+        Code taken from DialogueKit's DialogueConnector."""
+        if len(self.dialogue_history.utterances) == 0:
+            return
+
+        file_name = (
+            f"{_DIALOGUE_EXPORT_PATH}/"
+            f"{self.agent.id}_{self.user_simulator.id}.json"
+        )
+        json_file = []
+
+        # Check directory and read if exists.
+        if not os.path.exists(_DIALOGUE_EXPORT_PATH):
+            os.makedirs(_DIALOGUE_EXPORT_PATH)
+        if os.path.exists(file_name):
+            with open(file_name, encoding="utf-8") as json_file_out:
+                json_file = json.load(json_file_out)
+
+        dialogue_as_dict = self.dialogue_history.to_dict()
+        dialogue_as_dict["agent"] = self.agent.to_dict()
+        dialogue_as_dict["user"] = self.user_simulator.to_dict()
+
+        json_file.append(dialogue_as_dict)
+
+        with open(file_name, "w", encoding="utf-8") as outfile:
+            json.dump(json_file, outfile, indent=4)
+
+
+gym.register(
```

Comment:
Should this be under `if __name__ == "__main__":`?

Reply:
Done.